src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+    const gl = device.gl;
+    const level = calcLevels(target.width, target.height) - calcLevels(mipmap.width, mipmap.height);
+    const data = allocData(mipmap);
+
+    for (let f = 0; f < (target.cubemap ? 6 : 1); f++) {
+        const glTarget = target.cubemap ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + f : gl.TEXTURE_2D;
+
+        // read data back to CPU
+        readPixels(mipmap, f, data);
+
+        device.setTexture(target, 0);
+        gl.texImage2D(
+            glTarget,
+            level,
+            target._glInternalFormat,
+            mipmap.width,
+            mipmap.height,
+            0,
+            mipmap._glFormat,
+            mipmap._glPixelType,
+            data
+        );
+    }
+};
+
+// copy mipmap texture into one of the target's mipmaps (level is calculated from texture sizes)
+// target and mipmap can be 2d or cubemap textures
+const copyMipmap = (target, mipmap) => {
+    // copying directly from framebuffer to texture results in safari crashing
+    // so for now we use the slow version.
+    return copyMipmapSlow(target, mipmap);
+
+    // this version should be much faster, but crashes safari
+    // const device = target.device;
+    // const gl = device.gl;
+    // const level = calcLevels(target.width, target.height) - calcLevels(mipmap.width, mipmap.height);
+
+    // const oldRt = device.renderTarget;
+
+    // for (let f = 0; f < (target.cubemap ? 6 : 1); f++) {
+    //     const glTarget = target.cubemap ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + f : gl.TEXTURE_2D;
+    //     const renderTarget = new RenderTarget({
+    //         colorBuffer: mipmap,
+    //         face: f,
+    //         depth: false
+    //     });
+
+    //     // get the device to create gl interfaces
+    //     device.setRenderTarget(renderTarget);
+    //     device.updateBegin();
+
+    //     device.setTexture(target, 0);
+
+    //     // create the target mipmap level
+    //     gl.texImage2D(
+    //         glTarget,
+    //         level,
+    //         mipmap._glInternalFormat,
+    //         mipmap.width,
+    //         mipmap.height,
+    //         0,
+    //         mipmap._glFormat,
+    //         mipmap._glPixelType,
+    //         null
+    //     );
+
+    //     // copy it
+    //     gl.copyTexImage2D(
+    //         glTarget,
+    //         level,
+    //         mipmap._glInternalFormat,
+    //         0, 0, mipmap.width, mipmap.height,
+    //         0
+    //     );
+
+    //     device.updateEnd();
+    //     renderTarget.destroy();
+    // }
+
+    // // restore render target
+    // device.setRenderTarget(oldRt);
+    // device.updateBegin();
+};
+
+// generate mipmaps for the given target texture
+// target is either 2d equirect or cubemap with mipmaps = false
+const generateMipmaps = (target) => {
+    const device = target.device;
+    const gl = device.gl;
+    const numLevels = calcLevels(target.width, target.height);
+
+    let prevLevel = null;
+    for (let i = 1; i < numLevels; ++i) {
```

Comment:
this could be worth splitting into 3 loops .. first to allocate textures, then to render all, and last to delete resources. Otherwise it's possible a texture creation will need to block CPU and wait for previous rendering to finish.

Reply:
I considered adding a texture cache. Perhaps that with another function to clear the cache?

Reply:
yep .. or just a local array? Or is it likely we'd get some texture re-use from the cache?